src/joining_type.rs

```diff
@@ -0,0 +1,82 @@
+use std::collections::{BTreeMap, BTreeSet};
+
+use ucd_parse::{self, ArabicShaping};
+
+use args::ArgMatches;
+use error::{Error, Result};
+use general_category;
+use util::PropertyValues;
+
+pub fn command(args: ArgMatches) -> Result<()> {
+    let dir = args.ucd_dir()?;
+    let propvals = PropertyValues::from_ucd_dir(&dir)?;
+    let rows: Vec<ArabicShaping> = ucd_parse::parse(&dir)?;
+    let unexpanded_gc = ucd_parse::parse(&dir)?;
+    let gc =
+        general_category::expand_into_categories(unexpanded_gc, &propvals)?;
+
+    // Collect each joining type into an ordered set.
+    let mut by_type: BTreeMap<String, BTreeSet<u32>> = BTreeMap::new();
+    let mut assigned = BTreeSet::new();
+    for row in rows {
+        assigned.insert(row.codepoint.value());
+        let jt =
+            propvals.canonical("jt", row.joining_type.as_str())?.to_string();
+        by_type
+            .entry(jt)
+            .or_insert(BTreeSet::new())
+            .insert(row.codepoint.value());
+    }
+    // Process the codepoints that are not listed as per the note in
+    // ArabicShaping.txt:
+    //
+    // Note: Code points that are not explicitly listed in this file are either
+    // of joining type T or U:
+    //
+    // - Those that are not explicitly listed and that are of General Category
+    //   Mn, Me, or Cf have joining type T.
+    // - All others not explicitly listed have joining type U.
+    let transparent_name = propvals.canonical("jt", "transparent")?;
+    let non_joining_name = propvals.canonical("jt", "non_joining")?;
+    let transparent_categories = ["Mn", "Me", "Cf"]
+        .iter()
+        .map(|cat| {
+            propvals.canonical("gc", cat).and_then(|name| {
+                gc.get(&name).ok_or_else(|| {
+                    Error::Other(format!(
+                        "Unable to find general category '{}'",
+                        name
+                    ))
```

Comment:
I would be inclined to panic here, no? That is, this error should never be reached. The only way it could be reached, I think, is with a programmer error.

Reply:
Yes you're right. An invalid value needs to make it though canonicalization successfully too before hitting this branch. I've changed it.